pkg/idutil/imagewalker/imagewalker.go

```diff
@@ -98,6 +118,101 @@ func (w *ImageWalker) Walk(ctx context.Context, req string) (int, error) {
 	return matchCount, nil
 }
 
+// WalkCriRm walks images and calls w.OnFoundCriRm .
+// Only effective when in the k8s.io namespace and kube-hide-dupe is enabled.
+// The WalkCriRm deletes non-repo:tag items such as repo:digest when in the no-other-repo:tag scenario.
+func (w *ImageWalker) WalkCriRm(ctx context.Context, req string) (int, error) {
+	var filters []string
+	var parsedReferenceStr, repo string
+	var imageTag, imagesRepo []images.Image
+	var tagNum int
+
+	parsedReference, err := referenceutil.Parse(req)
+	if err == nil {
+		parsedReferenceStr = parsedReference.String()
+		filters = append(filters, fmt.Sprintf("name==%s", parsedReferenceStr))
+	}
+	//Get the image ID , if reg == imageTag use
+	image, err := w.Client.GetImage(ctx, parsedReferenceStr)
+	if err != nil {
+		repo = req
+	} else {
+		repo = strings.Split(image.Target().Digest.String(), ":")[1][:12]
+	}
+
+	filters = append(filters,
+		fmt.Sprintf("name==%s", req),
+		fmt.Sprintf("target.digest~=^sha256:%s.*$", regexp.QuoteMeta(repo)),
+		fmt.Sprintf("target.digest~=^%s.*$", regexp.QuoteMeta(repo)),
+	)
+
+	images, err := w.Client.ImageService().List(ctx, filters...)
+	if err != nil {
+		return -1, err
+	}
+
+	// to handle the `rmi -f` case where returned images are different but
+	// have the same short prefix.
+	uniqueImages := make(map[digest.Digest]bool)
+	nameMatchIndex := -1
+
+	//Distinguish between tag and non-tag
+	for _, img := range images {
+		ref := img.Name
+		parsed, err := reference.ParseAnyReference(ref)
+		if err != nil {
+			continue
+		}
+		if _, ok := parsed.(reference.Tagged); ok {
```

Comment:
https://github.com/containerd/nerdctl/pull/3744#discussion_r1894316679 @apostasie
Here, I need to obtain the interface corresponding to the image name, and then use interface.type to distinguish whether it's a standard tag( reference.Tagged) or a non-standard tag().
I simply expected standard implementation, but referenceutil.Parse works too.

Reply:
@fengwei0328 nerdctl does support additional image reference names that are not supported by the standard distribution library (ipfs and oci-archive custom schemes, ipfs ids).

That is the reason we should use nerdctl Parse instead.

Reply:
Thanks a lot for the changes @fengwei0328 
Let me do another review real quick.